### PR TITLE
Searcher warming API, take 2

### PR DESCRIPTION
**Pull request by shikhar**
`Warmer` implementations can be registered with the `IndexReaderBuilder` and used to maintain any segment-level caches effectively.

By integrating `Warmer::warm()` with the reload mechanism, we can ensure that any new segments are warmed before they can be exercised via a new generation of `Searcher`s.

`Warmer::garbage_collect()` is called when some searcher generation has been fully dropped, with the set of segment IDs that are still active. This allows implementations to prune redundant state.

### CODE REVIEW

**Comment by shikhar** [comment]
hello, this should be ready for review now!

File: src/reader/mod.rs
Comment (fulmicoton) [inline src/reader/mod.rs]:
```suggestion
    /// See [Self::searcher()].
```
This might seem like a weird suggestion. The documentation seemed to suggest that the called to `num_searchers()` would block.

File: src/reader/mod.rs
Comment (fulmicoton) [inline src/reader/mod.rs:124]:
```suggestion
    pub fn set_warmers(mut self, warmers: Vec<Weak<dyn Warmer>>) -> IndexReaderBuilder {
```

File: src/reader/warming.rs
Comment (fulmicoton) [inline src/reader/warming.rs]:
```suggestion
        if self.gc_thread.is_some() {
            return Ok(false);
        }
        let weak_inner = Arc::downgrade(this);
        let handle = std::thread::Builder::new()
            .name("tantivy-warm-gc".to_owned())
            .spawn(|| Self::gc_loop(weak_inner))
            .map_err(|_| {
               TantivyError::SystemError("Failed to spawn warmer GC thread".to_owned())
             })?;
        self.gc_thread = Some(handle);
        Ok(true)
```

Comment (fulmicoton) [reply src/reader/warming.rs]: Super minor. The code style in tantivy is quite a mess, but we are trying to move to early returns.

**Comment by fulmicoton** [comment]
@shikhar Great job. But before merging this, I'd like to discuss an alternate design to drop the need for a gc thread within tantivy, and possibly ease the user-side usage.

My understanding is that warmers compute some kind of byproduct associated to a searcher or a segment readers.
For instance, some statistics, or a cache, or whatever.

We cannot just assume that any searcher that we use is in sync with the last state of the warmer due. 
For instance, the following sequence of action is possible
- acquire searcher  gen N
- new generation of searcher. last warmer state is gen N+1
- actual use of searcher, using the warmer state

Your GC approach made it possible to handle the lifecycles of the warmer byproduct (which is great). 
However it does not help the user in a possession of a searcher to access the right warmer byproduct.

If the byproduct is actually a per-segment object it is possible to key these byproducts by segment id+delete opstamp.
If it is searcher level, the set of (segment_id + delete opstamp) and their is the right identifier.

This is possible. But this might be tricky.

As an alternative, should we could associate that byproduct directly with the Searcher with an Arc shared by all searcher?
`Drop` would take care of the cleaning up if needed (sometimes it might mean file deletion).
(Error handling would not be possible, but well it is not done either today).

That would solve:
- cleanup for things living in RAM
- the problem of accessing the right byproduct from the  searcher.

Of course the type system will not be nice with us...
But we could do some dynamic magic,  and use `Any` or even use a `typedmap`.

```
let my_cache: MyWarmer::ByProduct = searcher.warmer_byproduct(&warmer)
```

**Comment by shikhar** [comment]
Thanks for the detailed feedback @fulmicoton 

I see your point about deletes being the mutable layer on top of immutable segments, making it possibly incorrect to treat warmer byproducts as immutable for a given segment...

I believe this won't affect the `CachedQuery` that I was developing because it does not bother with the alive bitset in accumulating doc sets for caching by `SegmentId` (not exercising `Searcher` or `Collector`, iterating over underlying query's scorers directly [1]). AFAIU, [deletes would still be respected eventually using the live `Searcher` during collection](https://github.com/quickwit-inc/tantivy/blob/732f6847c0632bc086ea37f39d5f9d623aaa473d/src/collector/mod.rs#L181-L187).

The incremental indexing interval is sub-minute so I was hoping to maximize cache effectiveness by keeping it segment-level (keyed by `SegmentId`) rather than index-level (keyed by some `Searcher` generation token). There is a decent rate of clobbering old documents so delete opstamp in the key would probably cause similar churn.

Wondering if there is an API possibility that allows for safe segment-level caching. Would it suffice to warn users in the docs   about ignoring deletes for their byproducts?


[1]  ED: just checked and do need to pass `Searcher` to the underlying `query.weight()`..

**Comment by shikhar** [comment]
Maybe changing signature to

`warm(readers: &SegmentReader)`

where readers have been opened/cloned with empty `AliveSet`s

File: src/reader/warming.rs
Comment (shikhar) [inline src/reader/warming.rs]: Just came across the https://docs.rs/census/latest/census/ crate, which tantivy already uses...

**Comment by fulmicoton** [comment]
> I see your point about deletes being the mutable layer on top of immutable segments, making it possibly incorrect to treat warmer byproducts as immutable for a given segment...

And if the byproduct is a function not of a segment, but of the searcher entirely, we make it quite difficult to the user.

> The incremental indexing interval is sub-minute so I was hoping to maximize cache effectiveness by keeping it segment-level (keyed by SegmentId)

There are a lot of use cases for which segment-level byproducts is the most reasonable approach.
In your case, you can rely on the DeleteBitset to save the day and that's great. But someone who wants to pre-compute aggregates (average of price per product type for instance) would not be able to just key by segment.

I think we must leave it to the user to decide what they want to key their cache on.
This is not something I am disputing here.

Let's come back to your case...
You could keep warming up at the segment level.

In your case, the warmed-up by products would probably live in a cache (global or attached to your Warmer object) that goes :
```
HashMap<SegmentId, Weak<SegmentLevelCache>>
```

The warmer would first check if the byproduct is already cached.
If it is isn't, it would produce it and add it to the cache.
The byproduct itself would be an `HashMap<SegmentId, Arc<SegmentLevelCache>>` for all of the segments.

GC would be replaced by the drop of the last `Arc<SegmentLevelCache>` for a given `SegmentId`.

On consumption, you could get access to your data by requesting it from your `Searcher`, or through your cache (if you decided to make it global for instance).

**Comment by shikhar** [comment]
I made updates to identify the data accessed by searcher generations more accurately, using a map from `SegmentId -> NumDeletedDocs`. This is different from delete `Opstamp` but I think gives us the same effect while exposing a more meaningful value, since  number of deleted docs can only grow with new generations. Happy to change this to `Opstamp` if that seems preferable!

A use-case which is aggregating can retrieve `Searcher::generation()` and use that as a cache key at the index-level, or `(SegmentId, NumDeletedDocs)` at the segment-level.

And `SegmentId` is a reasonable cache key if storing document-level info - I was referring to this as segment-level before, but I think the distinguishing feature is it is at the `DocId` level and hence delete filtering during eventual collection can be effective.

---

I am still trying to understand your alternate API suggestion where a GC thread is not needed. Sounds like the idea is to move ownership of the byproducts to the `Searcher`. Would you mind explaining it again? 😅 

> The byproduct itself would be an HashMap<SegmentId, Arc<SegmentLevelCache>> for all of the segments.

`for all the searchers`, right?

p.s. learned about `typedmap` from your mention, pretty cool

**Comment by shikhar** [comment]
BTW to make the usage a bit more concrete, here are some of the types in my `CachedQuery` (still in flux)

```rust
type DataGeneration = u64;

// Implemented for `FunctionQuery` over a `ValueSource`, that can filter and optionally score using a Rust function over some per-doc value.
// `ValueSource` is implemented for a type providing mutable in-memory key:value map (also for `Field` assuming it's a fastfield...)
// the `ValueSource` generation is incremented when the data changes (bulk replacement to start with via files in object storage, but can also stream updates to a concurrent map)
// this allows joining against data which changes very frequently, ala Solr EFFs (https://solr.apache.org/guide/8_11/working-with-external-files-and-processes.html)
pub trait CacheableQuery: Query + Clone {
    fn generation(&self) -> DataGeneration;
}

enum Matches {
    Empty,
    Sparse(Arc<[(DocId, Score)]>),
    Dense(Arc<[Score]>),
    Const(Score, Arc<RoaringBitmap>),
}

type Cache = DashMap<SegmentId, (DataGeneration, Matches)>;

// scorer() returns a Matches-variant-specific `Scorer`
pub struct CachedQuery<Q>
where
    Q: CacheableQuery,
{
    query: Q,
    scoring_enabled: bool,
    cache: Arc<Cache>,
}

impl<Q> Warmer for CachedQuery<Q>
where
    Q: CacheableQuery
{
// Warming updates cache if a SegmentId is missing OR data generation for cached data is out-of-date.
// GC removes non-live SegmentId from cache
}

```

File: src/core/searcher.rs
Comment (fulmicoton) [inline src/core/searcher.rs]: Ah interesting. Technically this is correct because we cannot undelete a message, but I'd rather we used the `delete_opstamp` because this is what we use to identify a segment version in the code.
I'm ok to consider refactoring tantivy to `num_deleted_docs` though if we see some cool benefits.

That value is accessible from the `SegmentMeta`.

File: src/core/searcher.rs
Comment (fulmicoton) [inline src/core/searcher.rs:40]: Oh I knew cloned, but I didn't know about `.copied()`. Well now I am not sure of the benefits of using `.copied()` instead of `.cloned()`. It tells the reader it is "just a copy"?

File: src/core/searcher.rs
Comment (fulmicoton) [reply src/core/searcher.rs]: I think the change is simple, rename the function `from_segment_metas` and pass it SegmentMetas.

The call site already has them.

File: src/reader/mod.rs
Comment (fulmicoton) [inline src/reader/mod.rs]: searchable_segments above is your segment meta.

The reason for the weird lock is to make sure merge+GC does not remove files as we are in the middle of opening segments.

File: src/reader/warming.rs
Comment (fulmicoton) [inline src/reader/warming.rs:56]: I am confused. Why do we need `Vec<SearcherGenerationToken>` here?

Isn't `HashMap<SearcherIndexGeneration, SearcherGenerationToken>` suficient ?
I think all `SearcherGenerationToken` for a given SearcherIndexGeneration are identical `Weak` reference coupled to the same set of `Arc` clones.

Comment (fulmicoton) [reply src/reader/warming.rs:56]: If the above is correct, the HashMap is useless too...
It would become 

`Vec<SearcherGenerationToken>`

File: src/reader/warming.rs
Comment (fulmicoton) [inline src/reader/warming.rs]: Rename `warm_new_searcher_generation`

File: src/reader/warming.rs
Comment (fulmicoton) [inline src/reader/warming.rs]: Rename `warm_new_search_generation`

File: src/reader/warming.rs
Comment (fulmicoton) [inline src/reader/warming.rs]: Performance does not matter but well.... `swap_remove` is more indicated here..

**Review by fulmicoton — CHANGES_REQUESTED** [review]
See comments inline

**Comment by fulmicoton** [comment]
@shikhar  Let's forget about my suggestion and focus on landing your PR as is. I made a couple of comments however.

File: src/reader/warming.rs
Comment (shikhar) [reply src/reader/warming.rs:56]: > I think all `SearcherGenerationToken` for a given `SearcherIndexGeneration` are identical `Weak` reference coupled to the same set of Arc clones.

I believe `reload()` on an unchanged index can result in distinct sets of `Searcher` with the same underlying `SearcherIndexGeneration`?

File: src/reader/warming.rs
Comment (shikhar) [reply src/reader/warming.rs]: ooh, makes sense!

File: src/core/searcher.rs
Comment (shikhar) [reply src/core/searcher.rs:40]: Yes, I guess that's the only benefit since `Clone` will (unless manually implemented against expectations...) also just copy for a `Copy` type.

File: src/core/searcher.rs
Comment (shikhar) [reply src/core/searcher.rs]: > I'm ok to consider refactoring tantivy to num_deleted_docs though if we see some cool benefits.

Hmm, I think the only benefit is that it is a bit more meaningful in interpreting, opstamp feels opaque. Also `Option<u64>` vs `u32` - there is no need for optionality.

I changed it to the delete opstamp for consistency.

File: src/reader/mod.rs
Comment (shikhar) [reply src/reader/mod.rs]: Factored it out into new method `open()` that returns the generation & readers

File: src/core/searcher.rs
Comment (fulmicoton) [reply src/core/searcher.rs]: Yes that's interesting. I opened #1257.
I am not entirely sure there are no negative side effects, but on the paper that looks like a good idea.

File: src/reader/warming.rs
Comment (fulmicoton) [reply src/reader/warming.rs:56]: That's true. They will be identical though...

In your use case, your warmed data is keyed by `SegmentId` only and it might not really matter much... But let's dig a little bit.

One could imagine a Warmer that grabs updated values from an external source (say price from a SQL db), grabs the docid->someprimaryid from a fast field and creates a column `docid -> newprice using the `Warmer`. 
The Searcher is supposed to a consistent view of the index throughout its lifetime.

Calling reload manually, even without any change in the index is meaningful in that world: it updates the price.

So we want the API to allow:
- Warming stuff, and keeping things per `(Segment, DeleteOpstamp)`
- Warming stuff, and keeping things per `Searcher generation`.

**Comment by fulmicoton** [comment]
@shikhar I added one commit here:
https://github.com/quickwit-inc/tantivy/pull/1258/

    
- Using Inventory for the SearcherIndexGeneration
- Added an example.
- Added a searcher generation to warmer implementers to rely on that granularity
- Exporting census objects.
- Updated Changelog.

Can you have a look?

File: src/reader/warming.rs
Comment (shikhar) [reply src/reader/warming.rs:56]: > Calling reload manually, even without any change in the index is meaningful in that world

`warm()` will indeed get called -- it's a redundant `garbage_collect()` call that can be avoided when one of those `SearcherGenerationToken`s that point to different object but equal `SearcherIndexGeneration` are dropped. But I think not that big a deal to have a redundant call, as we always provide the set of live `SearcherIndexGeneration` to test state against.

**Comment by shikhar** [comment]
Superseded by #1258

**Pull request closed without merging**
